tests/test_jointcal.py

```diff
@@ -0,0 +1,87 @@
+# See COPYRIGHT file at the top of the source tree.
+import unittest
+import unittest.mock
+
+import lsst.log
+import lsst.utils
+
+import lsst.jointcal
+from lsst.jointcal import MinimizeResult
+import lsst.jointcal.chi2
+import lsst.jointcal.testUtils
+
+
+# for MemoryTestCase
+def setup_module(module):
+    lsst.utils.tests.init()
+
+
+class TestJointcalIterateFit(lsst.utils.tests.TestCase):
+    def setUp(self):
+        struct = lsst.jointcal.testUtils.createTwoFakeCcdImages(100, 100)
+        self.ccdImageList = struct.ccdImageList
+        # so that countStars() returns nonzero results
+        for ccdImage in self.ccdImageList:
+            ccdImage.resetCatalogForFit()
+
+        self.config = lsst.jointcal.jointcal.JointcalConfig()
+        # disable both, so it doesn't configure any refObjLoaders
+        self.config.doAstrometry = False
+        self.config.doPhotometry = False
+        self.jointcal = lsst.jointcal.JointcalTask(config=self.config)
+
+        self.goodChi2 = lsst.jointcal.chi2.Chi2Statistic()
+        self.goodChi2.chi2 = 200.0
+        self.goodChi2.ndof = 100
+
+        self.badChi2 = lsst.jointcal.chi2.Chi2Statistic()
+        self.badChi2.chi2 = 600.0
+        self.badChi2.ndof = 100
+
+        self.fitter = unittest.mock.Mock(spec=lsst.jointcal.PhotometryFit)
+        self.fitter.computeChi2.return_value = self.goodChi2
+        self.fitter.minimize.return_value = MinimizeResult.Converged
+        self.associations = unittest.mock.Mock(spec=lsst.jointcal.Associations)
+        self.associations.getCcdImageList.return_value = self.ccdImageList
+
+    def test_iterateFit_success(self):
+        chi2 = self.jointcal._iterate_fit(self.associations, self.fitter, 20, "testing", "")
+        self.assertEqual(chi2, self.goodChi2)
+        # Once for the for loop, the second time for the rank update.
+        self.assertEqual(self.fitter.minimize.call_count, 2)
+
+    def test_iterateFit_failed(self):
+        self.fitter.minimize.return_value = MinimizeResult.Failed
+
+        with self.assertRaises(RuntimeError):
+            self.jointcal._iterate_fit(self.associations, self.fitter, 20, "testing", "")
```

Comment:
How does this bit work? You tell python to raise a RuntimeError and then do some stuff? And because it fails, the call_count is only equal to 1 rather than 2?

Reply:
I set `minimize.return_value` to a value that should cause `_iterate_fit` to raise. If `minimize` fails, we raise immediately and there's no "final" `minimize` call (`jointcal.py:782`, if you care), so only one call to `minimize`.